### PR TITLE
WFLY-368: adding link refs to naming subsystem config "lookup" bindings, a...

### PR DESCRIPTION
**Pull request by emmartins**
...llowing app code to retrieve the lookup jndi name programatically, (only) through Context#lookupLink().

### CODE REVIEW

File: naming/src/main/java/org/jboss/as/naming/NamingStore.java [hunk 1]
```diff
@@ -86,6 +86,16 @@
...
 
     /**
+     * Retrieves the named object, following links except for the terminal atomic component of the name.
+     * If the object bound to <tt>name</tt> is not a link, returns the object itself.
+     *
+     * @param name The entry name
+     * @return The link from the store.
+     * @throws NamingException If any errors occur.
+     */
+    Object lookupLink(Name name) throws NamingException;
```
Comment (stuartwdouglas) [inline naming/src/main/java/org/jboss/as/naming/NamingStore.java]: I don't see why this needs a special method on the naming store, as the naming context already handles LinkRef lookup logic

File: naming/src/main/java/org/jboss/as/naming/ServiceBasedNamingStore.java [hunk 2]
```diff
@@ -255,6 +260,16 @@ public Object run() {
         return children;
     }
 
+    @Override
+    public Object lookupLink(Name name) throws NamingException {
+        Object object = lookup(name, false);
+        if (object instanceof LookupBindingManagedReferenceFactory) {
+            return LookupBindingManagedReferenceFactory.class.cast(object).getLink();
```
Comment (stuartwdouglas) [inline naming/src/main/java/org/jboss/as/naming/ServiceBasedNamingStore.java]: I don't like this at all. This should just work with LinkRef, if should not need a special ManagedReferenceFactory

File: naming/src/main/java/org/jboss/as/naming/subsystem/LookupBindingManagedReferenceFactory.java [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+import org.jboss.as.naming.ManagedReference;
+
+import javax.naming.InitialContext;
+import javax.naming.LinkRef;
+
+/**
+ * The {@link org.jboss.as.naming.ManagedReferenceFactory} for config's lookup bindings, which expose the source's lookup name.
+ * @author Eduardo Martins
+ */
+public class LookupBindingManagedReferenceFactory implements ContextListAndJndiViewManagedReferenceFactory {
```
Comment (stuartwdouglas) [inline naming/src/main/java/org/jboss/as/naming/subsystem/LookupBindingManagedReferenceFactory.java]: Why are we doing this? If you are going to use a LinkRef use a LinkRef, I really don't see the point of this weird hybrid

**Comment by stuartwdouglas** [comment]
I am rejecting this, as I don't want to add internal hacks to make lookup appear to use a LinkRef even though it doesn't really.

The original fix was rejected because Weld resource injection attempts to use the binder service directly in order to get more information about the injected value, I would rather just fix this code to be able to deal with links than add this sort of hack. 

In the JIRA David mentions that an alternative solution was agreed to (adding a link element that uses LinkRef), although if the Weld code can handle the links I don't really see any problem with just using a LinkRef directly.

**Comment by emmartins** [comment]
1. I tried linkrefs long ago and it broke a lot of stuff, the concrete issue is the injection of the binder service into another service, the link ref is not resolved at that level and the injection fails. I recall this was specially problematic on the tck, wrt datasources.

2. We also discussed that additional link binding type, beyond what's in the JIRA, but it was agreed that it was not critical and would probably be confusing.

3. I don't consider this solution an hack, IMHO it's just a different form of a link ref, that is not problematic wrt MSC service injection or anything else.

**Pull request closed without merging**